pkg/koordlet/resmanager/blkio_reconcile.go

```diff
@@ -0,0 +1,636 @@
+			if podVolumeResult != "" {
+				if podBlkIOQoS, err = parseBlkIOResult(podVolumeResult); err != nil {
+					klog.Errorf("%s: unmarshal pod annotation %v failed, error %v", BlkIOReconcileName, extension.AnnotationPodBlkioQoS, err)
+					continue
+				}
+			}
+		} else {
+			continue
+		}
+		klog.V(4).Infof("%s: start to reconcile pod %s/%s blkio config", BlkIOReconcileName, podMeta.Pod.Namespace, podMeta.Pod.Name)
+		err = b.updateBlkIOConfig(
+			podBlkIOQoS.Blocks,
+			podMeta,
+			blkioUpdater{
+				absolutePath:    util.GetPodCgroupBlkIOAbsolutePath(podMeta.CgroupDir),
+				dynamicPath:     podMeta.CgroupDir,
+				getDiskRecorder: getBlkIORecorder,
+				getUpdaterFunc:  getBlkIOUpdaterFromBlockCfg,
+				getRemoverFunc:  getBlkIORemoverFromDiskNumber,
+			},
+		)
+		if err != nil {
+			klog.Errorf("%s: fail to update pod %s/%s blkio config: %s", BlkIOReconcileName, podMeta.Pod.Namespace, podMeta.Pod.Name, err.Error())
+		} else {
+			klog.V(4).Infof("%s: reconcile pod %s/%s blkio config finished", BlkIOReconcileName, podMeta.Pod.Namespace, podMeta.Pod.Name)
+		}
+	}
+}
+
+type blkioUpdater struct {
+	dynamicPath  string
+	absolutePath string
+
+	getDiskRecorder GetDiskRecorderFunc
+	getUpdaterFunc  GetUpdaterFunc
+	getRemoverFunc  GetRemoverFunc
+}
+
+// update blkio cgroup files
+// podMeta == nil when BlockType is BlockTypeDevice or BlockTypeVolumeGroup
+// podMeta != nil when BlockType is BlockTypePodVolume
+func (b *BlkIOReconcile) updateBlkIOConfig(blocks []*slov1alpha1.BlockCfg, podMeta *statesinformer.PodMeta, blkioUpdater blkioUpdater) error {
+	if blkioUpdater.getDiskRecorder == nil {
+		return fmt.Errorf("getDiskRecorder can not be nil")
+	}
+	if blkioUpdater.getUpdaterFunc == nil || blkioUpdater.getRemoverFunc == nil {
+		return fmt.Errorf("getUpdaterFunc or getRemoverFunc can not be nil")
+	}
+	var resources []resourceexecutor.ResourceUpdater
+	diskConfigRecorder, err := blkioUpdater.getDiskRecorder(blkioUpdater.absolutePath)
+	if err != nil {
+		return fmt.Errorf("fail to get disk config recorder: %s", err.Error())
+	}
+	for _, block := range blocks {
+		diskNumber, err := b.getDiskNumberFromBlockCfg(block, podMeta)
+		if err != nil {
+			return fmt.Errorf("fail to get disk number from block %v: %s", block, err.Error())
+		}
+		diskConfigRecorder[diskNumber] = false
+		resources = append(resources, blkioUpdater.getUpdaterFunc(block, diskNumber, blkioUpdater.dynamicPath)...)
+	}
+	for diskNumber, needRemove := range diskConfigRecorder {
+		if needRemove {
+			resources = append(resources, blkioUpdater.getRemoverFunc(diskNumber, blkioUpdater.dynamicPath)...)
+		}
+	}
+	b.executor.UpdateBatch(true, resources...)
+	return nil
+}
+
+// deviceName: /dev/sdb
+// diskNumber: 253:16
+func (b *BlkIOReconcile) getDiskNumberFromDevice(deviceName string) (string, error) {
+	storageInfo, err := b.resmanager.metricCache.GetNodeLocalStorageInfo(&metriccache.QueryParam{})
```

Comment:
perhaps we only need to invoke `GetNodeLocalStorageInfo` once in a reconcile

Reply:
done